File: src/LaunchPad.Api/LocalDemo/DevPersonas.cs
using LaunchPad.Application.Common;

namespace LaunchPad.Api.LocalDemo;

public sealed record DevPersona(string Key, Guid EntraObjectId, string DisplayName, string Upn, string[] Roles);

/// <summary>
/// Fixed identities for local multi-role testing (see DevPersonaAuthHandler). Object ids
/// are constants so the seeded users in LocalDemoSeeder are the same people a tester
/// signs in as. Mirrored in src/LaunchPad.Web/src/dev/devPersonas.ts — keep the keys in step.
/// </summary>
public static class DevPersonas
{
    public static readonly DevPersona Ops = new("ops", Guid.Parse("00000000-0000-4000-8000-000000000001"), "Olivia Ops", "olivia.ops@example.com", [Roles.ProgramOps]);
    public static readonly DevPersona Exec = new("exec", Guid.Parse("00000000-0000-4000-8000-000000000002"), "Evan Exec", "evan.exec@example.com", [Roles.Executive]);
    public static readonly DevPersona HiringManager = new("hm", Guid.Parse("00000000-0000-4000-8000-000000000003"), "Harper Hiring", "harper.hiring@example.com", [Roles.HiringManager]);
    public static readonly DevPersona Sponsor = new("sponsor", Guid.Parse("00000000-0000-4000-8000-000000000010"), "Sam Sponsor", "sponsor.demo@example.com", [Roles.Sponsor]);
    public static readonly DevPersona Sponsor2 = new("sponsor2", Guid.Parse("00000000-0000-4000-8000-000000000011"), "Priya Shah", "priya.shah@example.com", [Roles.Sponsor]);
    public static readonly DevPersona SponsorNew = new("sponsor-new", Guid.Parse("00000000-0000-4000-8000-000000000012"), "Nia Newsponsor", "nia.newsponsor@example.com", [Roles.Sponsor]);
    public static readonly DevPersona Candidate1 = new("cand1", Guid.Parse("00000000-0000-4000-8000-000000000020"), "Jordan Rivera", "jordan.rivera@example.com", [Roles.Candidate]);
    public static readonly DevPersona Candidate2 = new("cand2", Guid.Parse("00000000-0000-4000-8000-000000000021"), "Casey Kim", "casey.kim@example.com", [Roles.Candidate]);
    public static readonly DevPersona Candidate3 = new("cand3", Guid.Parse("00000000-0000-4000-8000-000000000022"), "Morgan Lee", "morgan.lee@example.com", [Roles.Candidate]);
    public static readonly DevPersona CandidateCohort2 = new("cand-c2", Guid.Parse("00000000-0000-4000-8000-000000000023"), "Alex Torres", "alex.torres@example.com", [Roles.Candidate]);
    public static readonly DevPersona CandidateNew = new("cand-new", Guid.Parse("00000000-0000-4000-8000-000000000024"), "Riley Newcandidate", "riley.new@example.com", [Roles.Candidate]);
    public static readonly DevPersona OpsSponsor = new("ops-sponsor", Guid.Parse("00000000-0000-4000-8000-000000000030"), "Dana Dualrole", "dana.dual@example.com", [Roles.ProgramOps, Roles.Sponsor]);
    public static readonly DevPersona NoRole = new("norole", Guid.Parse("00000000-0000-4000-8000-000000000040"), "Nobody Noroles", "nobody@example.com", []);

    public static readonly IReadOnlyDictionary<string, DevPersona> ByKey = new[]
    {
        Ops, Exec, HiringManager, Sponsor, Sponsor2, SponsorNew,
        Candidate1, Candidate2, Candidate3, CandidateCohort2, CandidateNew, OpsSponsor, NoRole,
    }.ToDictionary(p => p.Key, StringComparer.Ordinal);
}

File: src/LaunchPad.Web/src/dev/devPersonas.ts
import { AppRoles, type AppRole } from '../auth/roles';
import { isMockMode } from './mockMode';

/**
 * Local-only multi-role testing (see LocalDemo/DevPersonaAuthHandler.cs in the API).
 * Same compile-time gating as mockMode.ts: import.meta.env.DEV is false in every
 * production build, so this is dead code there. Mutually exclusive with mock mode —
 * persona mode talks to the real local API, mock mode never does.
 *
 * The chosen persona lives in sessionStorage, which is per tab: each browser tab is an
 * independent signed-in user, which is what lets one browser drive a Sponsor and an
 * Ops approver side by side.
 */
export const isPersonaMode =
  import.meta.env.DEV && import.meta.env.VITE_DEV_PERSONAS === 'true' && !isMockMode;

export const PERSONA_HEADER = 'X-Dev-Persona';
const STORAGE_KEY = 'launchpad.devPersona';

export type DevPersona = { key: string; displayName: string; roles: AppRole[]; note: string };

// Mirrors src/LaunchPad.Api/LocalDemo/DevPersonas.cs — the API is the authority on
// roles; this copy only labels the picker and shapes navigation.
export const DEV_PERSONAS: DevPersona[] = [
  { key: 'ops', displayName: 'Olivia Ops', roles: [AppRoles.ProgramOps], note: 'Program Ops' },
  { key: 'exec', displayName: 'Evan Exec', roles: [AppRoles.Executive], note: 'Executive' },
  { key: 'hm', displayName: 'Harper Hiring', roles: [AppRoles.HiringManager], note: 'Hiring Manager' },
  { key: 'sponsor', displayName: 'Sam Sponsor', roles: [AppRoles.Sponsor], note: 'Sponsor, owns the demo projects' },
  { key: 'sponsor2', displayName: 'Priya Shah', roles: [AppRoles.Sponsor], note: 'Sponsor, a different organization' },
  { key: 'sponsor-new', displayName: 'Nia Newsponsor', roles: [AppRoles.Sponsor], note: 'Sponsor with no profile yet' },
  { key: 'cand1', displayName: 'Jordan Rivera', roles: [AppRoles.Candidate], note: 'Candidate, spring cohort' },
  { key: 'cand2', displayName: 'Casey Kim', roles: [AppRoles.Candidate], note: 'Candidate, spring cohort' },
  { key: 'cand3', displayName: 'Morgan Lee', roles: [AppRoles.Candidate], note: 'Candidate, spring cohort' },
  { key: 'cand-c2', displayName: 'Alex Torres', roles: [AppRoles.Candidate], note: 'Candidate, fall cohort' },
  { key: 'cand-new', displayName: 'Riley Newcandidate', roles: [AppRoles.Candidate], note: 'Candidate with no profile yet' },
  { key: 'ops-sponsor', displayName: 'Dana Dualrole', roles: [AppRoles.ProgramOps, AppRoles.Sponsor], note: 'Program Ops and Sponsor' },
  { key: 'norole', displayName: 'Nobody Noroles', roles: [], note: 'Signed in with no roles' },
];

export function getActivePersona(): DevPersona | null {
  if (!isPersonaMode) return null;
  try {
    const key = sessionStorage.getItem(STORAGE_KEY);
    return DEV_PERSONAS.find((p) => p.key === key) ?? null;
  } catch {
    return null;
  }
}

export function setActivePersona(key: string | null) {
  try {
    if (key) sessionStorage.setItem(STORAGE_KEY, key);
    else sessionStorage.removeItem(STORAGE_KEY);
  } catch {
    // Storage blocked: the picker simply shows again on reload.
  }
  // A full reload, not a state update: every cached query belongs to the previous
  // identity, and a reload is the one way to guarantee none of it survives.
  window.location.assign('/');
}
